src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -967,45 +968,105 @@ protected static UnmanagedCallersOnlyAttributeData DecodeUnmanagedCallersOnlyAtt
             {
                 foreach (var (key, value) in attribute.CommonNamedArguments)
                 {
-                    if (!UnmanagedCallersOnlyAttributeData.IsCallConvsTypedConstant(key, in value))
-                    {
-                        continue;
-                    }
+                    var namedArgumentDecoded = TryDecodeUnmanagedCallersOnlyCallConvsProperty(key, value, location, diagnostics);
 
-                    if (callingConventionTypes != null)
+                    if (namedArgumentDecoded.IsCallConvs)
                     {
-                        isValid = false;
+                        isValid = isValid && namedArgumentDecoded.IsValid;
+                        callingConventionTypes = namedArgumentDecoded.CallConvs;
                     }
+                }
+            }
 
-                    if (value.Values.IsDefaultOrEmpty)
-                    {
-                        callingConventionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
-                        continue;
-                    }
+            return UnmanagedCallersOnlyAttributeData.Create(callingConventionTypes, isValid);
+        }
 
-                    var builder = PooledHashSet<INamedTypeSymbolInternal>.GetInstance();
-                    foreach (var callConvTypedConstant in value.Values)
-                    {
-                        Debug.Assert(callConvTypedConstant.Kind == TypedConstantKind.Type);
-                        if (!(callConvTypedConstant.ValueInternal is NamedTypeSymbol callConvType)
-                            || !FunctionPointerTypeSymbol.IsCallingConventionModifier(callConvType))
-                        {
-                            // `{0}` is not a valid calling convention type for 'UnmanagedCallersOnly'.
-                            diagnostics?.Add(ErrorCode.ERR_InvalidUnmanagedCallersOnlyCallConv, location!, callConvTypedConstant.ValueInternal ?? "null");
-                            isValid = false;
-                        }
-                        else
-                        {
-                            _ = builder.Add(callConvType);
-                        }
+        internal static (bool IsCallConvs, ImmutableHashSet<INamedTypeSymbolInternal>? CallConvs, bool IsValid) TryDecodeUnmanagedCallersOnlyCallConvsProperty(
+            string key,
+            TypedConstant value,
+            Location? location,
+            DiagnosticBag? diagnostics)
+        {
+            ImmutableHashSet<INamedTypeSymbolInternal>? callingConventionTypes = null;
+            bool isValid = true;
 
-                    }
-                    callingConventionTypes = builder.ToImmutableHashSet();
-                    builder.Free();
+            if (!UnmanagedCallersOnlyAttributeData.IsCallConvsTypedConstant(key, in value))
```

Comment:
>key [](start = 76, length = 3)

Where do we verify that this refers to a property?

Reply:
I need to rename this, it's actually a field. I'll update.

---
In reply to: [493035146](https://github.com/dotnet/roslyn/pull/47950#discussion_r493035146) [](ancestors = 493035146)

Reply:
When decoding from source, we our existing `NamedArguments` infrastructure does not keep track of whether it's a field or a property, so I cannot verify that. I can do so when decoding from IL though.

---
In reply to: [493878851](https://github.com/dotnet/roslyn/pull/47950#discussion_r493878851) [](ancestors = 493878851,493035146)

Reply:
> When decoding from source, we our existing NamedArguments infrastructure does not keep track of whether it's a field or a property, so I cannot verify that. I can do so when decoding from IL though.

We need to check this from source as well. See how we are doing similar check for Obsolete attribute in AttributeData.DecodeObsoleteAttribute

---
In reply to: [493880852](https://github.com/dotnet/roslyn/pull/47950#discussion_r493880852) [](ancestors = 493880852,493878851,493035146)